framework/src/relationshipmanagers/RedistributeProperties.C

```diff
@@ -0,0 +1,212 @@
+
+void
+RedistributeProperties::operator()(const MeshBase::const_element_iterator &,
+                                   const MeshBase::const_element_iterator &,
+                                   processor_id_type,
+                                   map_type &)
+{
+}
+
+std::unique_ptr<GhostingFunctor>
+RedistributeProperties::clone() const
+{
+  return std::make_unique<RedistributeProperties>(*this);
+}
+
+std::string
+RedistributeProperties::getInfo() const
+{
+  return "RedistributeProperties";
+}
+
+// the LHS ("this" object) in MooseApp::addRelationshipManager is the existing RelationshipManager
+// object to which we are comparing the rhs to determine whether it should get added
+bool
+RedistributeProperties::operator>=(const RelationshipManager & rhs) const
+{
+  const auto * rm = dynamic_cast<const RedistributeProperties *>(&rhs);
+
+  // All RedistributeProperties objects are effectively equivalent
+  return rm;
+}
+
+void
+RedistributeProperties::addMaterialPropertyStorage(
+    std::vector<std::shared_ptr<MaterialData>> & mat_data, MaterialPropertyStorage & mat_props)
+{
+  _materials.emplace_back(&mat_data, &mat_props);
+}
+
+void
+RedistributeProperties::redistribute()
+{
+  const MeshBase & mesh = _moose_mesh->getMesh();
+  const processor_id_type pid = mesh.processor_id();
+
+  for (auto & [mat_data, mat_prop_store] : _materials)
+  {
+    // Once we've redistributed data elsewhere we'll delete it here.
+    //
+    // We don't need stored properties on non-local elements, and so
+    // we don't compute them and we don't *bother deleting them when
+    // non-local elements are coarsened away*, so this may be our last
+    // chance to avoid dangling pointers in future redistribute()
+    // calls.
+    std::set<const Elem *> elems_being_migrated;
+
+    if (mat_prop_store->hasStatefulProperties())
+    {
+      mooseAssert(
+          !Threads::in_threads,
+          "This routine has not been implemented for threads. Please query this routine before "
+          "a threaded region or contact a MOOSE developer to discuss.");
+      MaterialData & my_mat_data = *((*mat_data)[/*_tid*/ 0]); // Not threaded
+
+      std::array<MaterialPropertyStorage::PropsType *, 3> props_maps{
+          mat_prop_store->_props_elem.get(),
+          mat_prop_store->_props_elem_old.get(),
+          mat_prop_store->_props_elem_older.get()};
+
+      for (auto * const props_map_ptr : props_maps)
+      {
+        MaterialPropertyStorage::PropsType & props_map = *props_map_ptr;
+        typedef std::unordered_map<unsigned int, std::string> stored_props_type;
+        typedef std::tuple<stored_props_type, dof_id_type, int> stored_elem_type;
+        std::map<processor_id_type, std::vector<stored_elem_type>> props_to_push;
+
+        // Take non-const references here for compatibility with dataStore(T&)
+        for (auto & [elem, elem_props_map] : props_map)
+        {
+          // There better be an element here
+          libmesh_assert(elem);
```

Comment:
I prefer mooseAssert for future reference. No message is not super friendly to the next person hitting this

Reply:
I dont think it's because of a policy / difficulty to add a string to mooseAssert that we have less asserts than libmesh does. 

My guess would be because expertise about what could go wrong is much stronger within the libmesh devs than the MOOSE ones.